src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CaseReferenceValidator.java

```diff
@@ -0,0 +1,38 @@
+package uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd;
+
+import com.google.common.base.Strings;
+import io.vavr.control.Validation;
+import uk.gov.hmcts.reform.ccd.client.model.CaseDetails;
+
+import java.util.Optional;
+import javax.annotation.Nonnull;
+
+import static io.vavr.control.Validation.invalid;
+import static java.lang.String.format;
+
+class CaseReferenceValidator {
+    private static final String ATTACH_TO_CASE_REFERENCE = "attachToCaseReference";
+
+    @Nonnull
+    Validation<String, String> validate(CaseDetails theCase) {
+        return Optional.ofNullable(theCase)
+            .map(CaseDetails::getData)
+            .map(data -> data.get(ATTACH_TO_CASE_REFERENCE))
+            .flatMap(this::validateCaseRef)
+            .orElseGet(() -> invalid("No case reference supplied"));
+    }
+
+    private Optional<Validation<String, String>> validateCaseRef(Object caseRef) {
+        Optional<Validation<String, String>> valid = Optional.of(caseRef)
+            .filter(aRef -> aRef instanceof String)
+            .map(aRef -> ((String) aRef).replaceAll("[^0-9]", ""))
+            .map(Strings::emptyToNull)
+            .map(Validation::valid);
+        //This is done because java 8 does not have "Optional<T> orElse( Optional<T>)"
+        if (valid.isPresent()) {
+            return valid;
+        } else {
+            return Optional.of(invalid(format("Invalid case reference: '%s'", caseRef)));
+        }
```

Comment:
do we need to return optional at all here? looks like we could just `.orElse` and get rid of another flatmap one method above. plus second `.orElseGet` over there is not needed either as this method is always returning non empty optional value

Reply:
nope ... thats the problem with java 8 optional.
This is pretty  much how I have to do it.
give it a go if you like but this is the exact reason orElse() was added to optional in java 9

Reply:
I get what you mean. In short, question is "do we need `Optional<?>` as a returned value here?"